### PR TITLE
feat: grouped aggregates for low-cardinality columns

### PR DESCRIPTION
**Pull request by e-dard**
TLDR; this PR adds a method for doing grouped by aggregates on low-cardinality columns that is ~invariant to the number of rows being aggregated across. It could be useful for queries across large numbers of rows where you only want to group on lower cardinality columns.

## Background

This PR adds some initial support for `read_group` aggregates in Segment Store.
Currently it only supports one particular type of grouped aggregates; I will continue to add support for other cases until all general cases are covered.

Expressed in SQL, the kinds of read_group aggregate support (pinching a bit of InfluxDB terminology here...) added in this PR could be expressed in SQL like:

```sql
SELECT SUM(duration), COUNT(foo), MIN(bar) ...
FROM "cpu"
GROUP BY "region", "host", .... 
```

Concretely, for the code-paths in this PR to be hit the following conditions need to be true:

* No predicates on the query that apply to the segment (including time);
* All grouping columns are `RLEDictionary` encoded;

There can be predicates on the query, but they must not filter any data in the segment, so a time-range that overlaps the entire segment is fine, as is a filter that matches all values in a column.
The reason for these restrictions is that the algorithm that does the grouping leverages some pre-computed set data that all rle-encoded columns hold.

Typically grouped aggregates are done in a few steps:

1) Apply all filters and determine what set of rows you're working with on each column;
2) With this intermediate row set calculate each group key across all grouping columns for each `row_id` in the row set.
3) Store these group keys as keys to a hashmap and store the aggregates as the values, which can be built as you go.

An alternative approach is to not build a hashmap but instead to create another table view with the filtered rows that is then sorted on the grouping columns.
You can then stream through the columns you're aggregating in row-order and build the aggregates that way; the hashtable approach is often faster but YMMV.

However, for these scenarios I wanted to see if I could leverage the fact that we already had all the groups pre-computed on each column in compressed bitsets. 

## Implementation

Each rle-encoded columns stores a mapping between each `encoded_id` (the integer that encodes the logical string value in a row) and all of the `row_ids` that contain that value.
These are stored in compressed roaring bitmaps.

For example, suppose you have the following column `region`:

```
["east", "east", "west", "east", "south", "south"]
```

we may encode that as:

```
"east" -> 1
"south" -> 2
"west" -> 3
```

and then we will store the pre-computed rows in compressed bitsets for each _encoded_ value. 
Like this:

```
1: {0, 1, 3}
2: {4, 5}
3: {2}
```

where the sets are themselves compressed.

Now, given another column `az` with a pre-computed grouping like:

```
["az-b", "az-a", "az-e", "az-a", "az-a", "az-b"]

"az-a" -> 1
"az-b" -> 2
"az-e" -> 3

1: {1, 3, 4}
2: {0, 5}
3: {2}
```

we could simply intersect the bitsets for each group combination resulting in the final group keys and row_ids to apply the aggregates to:

```
(1, 1) -> {1, 3}
(1, 2) -> {0}
(2, 2) -> {5}
(3, 3) -> {2}

materialised, the groupings have the following rows.

(east, az-a) -> {1, 3}
(east, az-b) -> {0}
(south, az-b) -> {5}
(west, az-e) -> {2}
```

you can group by as many columns as needed this way as a chain of set intersections.

Cool. All good then?
Well not really because you can't tell what the group keys are unless you check all the candidate's sets against each other, so if you do it naively (which is what I have done here for a starter implementation), you very likely have a non-linear algorithm :-1:.

I think the approach is interesting through and I'm pretty sure you could implement a linear version by "draining" the bitsets as you calculate group keys, but I didn't continue down that road as there is other `read_group` surface area to cover first.

## Commits

Some of the less boring commits in the PR:

* 7bc9738 - This adds the basic algorithm for applying read_group aggregates to columns that support returning pre-computed bit-sets for each distinct value in the column. I think the algorithm is linearithmic (maybe quadratic) because it lazily generates all possible group keys via an cartesian product iterator and then does set operations to determine if each candidate group key is a valid group key. It does short-circuit aggressively though.
* 42f56f4 - Benchmarks used to assess performance. I'll describe these below.
* 7ae27f5 - With benchmarks comes profiling, and with profiling usually comes easy wins. This one checks whether any rows could exist for a group key before allocating a new bitset.
* 3f8208d - This one avoids copying all `encoded_ids` out of the column initially and defers the copy until we know we need to use it to materialise a logical column value.
* 1043495 - This one was fun. Profiling showed that a lot of time was being spent fetching the bitsets for each distinct column value out of their respective `BTreeMaps`. The maps are indexed by integers (the `encoded_id`s for the logical column values). Not only are the maps indexed by integers but they're also guaranteed to be `[0, N)` where `N` is the number of distinct values in the column. So let's use a vector here and simply index into it with the `encoded_id` integer. This was a big performance win.
* f524240 - Solidify the vector idea in the general encoding APIs.

## Results

There are three different benchmarks in the PR, which use a basic data-set that generates column data similar to that found in a tracing use-case.
There are low cardinality columns in the segment, e.g., things like `env`, `data_centre` or `cluster`. 
And there are high cardinality columns like `request_id`, `trace_id` and `span_id`. 
The code described in this PR only applies to grouping on low-cardinality columns that are RLE encoded.

The three benchmarks are as follows:

* One of them varies the cardinality of the results whilst fixing the number of rows in the segment to 500,000. The cardinality of the results just means the number of grouped aggregates produced (the number of group keys). It varies between 2, 20, 200 and 2,000.
* One of them fixes the cardinality (to 200), but varies the number of columns grouped to produce that result set. It varies grouping on 1, 2 or 3 columns.
* One of them fixes cardinality and columns and varies the number of rows in the segment.

The initial results seem interesting (to me anyway). 
Firstly, let's address the current problem which is above-linear behaviour of the current implementation.
Here is a graph showing this behaviour, taken from the first benchmark that varies cardinality for a 500,000 rows segment.

Screenshot 2020-12-04 at 21 18 29

Grouping throughput goes from `~250us` to calculate 200 grouped aggregates (`~800K` groups/second) to `~15ms` to calculate 2,000 grouped aggregates (`~130K).
Ideally we would expect the throughput to be the same as we increased the number of groups.

It's not all bad though.
One interesting aspect to this approach is that it's relatively immune to the actual size of the segments. 
That is, larger segments (number of rows) don't really change the performance of the grouping much.
Here are some benchmarks showing this:

Screenshot 2020-12-04 at 20 14 28

The benchmarks do the equivalent of `SELECT COUNT(duration) FROM cpu GROUP BY "data_centre", "cluster"` but varying the number of rows to process from 250,000 to 1 million.
You can see there is barely any performance change of the query between processing 250K rows (`~245 us`) and processing 1M rows (`~302 us`), which is pretty cool.

Finally, here is a cool [criterion](https://github.com/bheisler/criterion.rs) feature that shows the efforts of some performance tuning I did in the latter commits after doing some profiling:

Screenshot 2020-12-04 at 20 29 03

## Summary

I think for the right kind of queries where you want to get aggregates across low cardinality columns, this implementation could work really well.
One example would be when you have lots of potentially high cardinality data, but you want summary statistics over some of the lower cardinality columns, e.g.,

```sql
// show me the number of traces collected in each region and cluster, as well as the fastest and slowest span.

SELECT COUNT(trace_id), MIN(duration), MAX(duration)
FROM traces
GROUP BY "aws-region", "cluster-type"
```

Even if there are millions of traces in the segment this query will still perform well.

For higher cardinality columns however it will probably be better to stick to traditional hashmap/sorting approaches.

## Next Steps

* I'm going to move onto implementing more general `read_group` support with predicates and higher cardinality columns.
* I think it would be worth figuring out where the overlap is between using this code-path and using a more general hashmap based approach, and then switching between them in IOx at runtime.
* Figure out a better algorithm than testing every group key combination in this PR.

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
I can't say I fully groked this entire PR , but I did read it pretty carefully and I think it is good to merge to me.

I was  somewhat confused about the explanation in the PR's description -- 
In your example, I can't tell if we are grouping by "region", or "region1", "region2" (aka are the two encoded columns the *same* column that needs to be aggregated, or are they two columns that form a compound group key-- I think it is the latter , and I found https://github.com/influxdata/influxdb_iox/pull/527/files#diff-05b254f51e0af46c018ef89da62237a52959aed4c59c6651b4f83835295fe51dR383-R413 to help understanding what is going on with this code better

**Review by alamb — APPROVED** [review]

File: segment_store/src/segment.rs [hunk 1]
```diff
@@ -323,13 +323,189 @@ impl Segment {
...
+        let all_group_cols_pre_computed = group_columns.iter().all(|col_name| {
+            self.column_by_name(col_name)
+                .properties()
+                .has_pre_computed_row_ids
+        });
+        if predicates.is_empty() && all_group_cols_pre_computed {
+            return self.read_group_all_rows_all_rle(group_columns, aggregates);
+        }
+
+        ReadGroupResult::default()
```
[the remark this answers is not in the document]
Comment (e-dard) [reply segment_store/src/segment.rs:349]: The general case is coming next. It won't be `default()` for long.

File: segment_store/src/column/dictionary.rs [hunk 2]
```diff
@@ -735,11 +735,61 @@ mod test {
...
+
+        enc.push_additional(Some("east".to_string()), 4); // 0, 1, 2, 3
+        enc.push_additional(Some("west".to_string()), 2); // 4, 5
+        enc.push_none(); // 6
+        enc.push_additional(Some("zoo".to_string()), 1); // 7
+
+        match enc {
+            Encoding::RLE(_) => {
+                assert_eq!(
+                    enc.group_row_ids().unwrap_left(),
```
[the remark this answers is not in the document]
Comment (e-dard) [reply segment_store/src/column/dictionary.rs:761]: They don't return the same value. One returns a `Vec<&T>` and one returns a `Vec<T>`. Since the types are both owned, but one has a borrowed  inner type `Cow` doesn't work, so `Either` is just a general purpose enum that's useful for internal APIs.

File: segment_store/src/table.rs [hunk 3]
```diff
@@ -164,16 +174,41 @@ impl Table {
...
-        group_columns: Vec<ColumnName<'a>>,
-        aggregates: Vec<(ColumnName<'a>, AggregateType)>,
-    ) -> BTreeMap<GroupKey, Vec<(ColumnName<'a>, AggregateResult<'_>)>> {
+        &'a self,
+        predicates: &[Predicate<'a>],
+        group_columns: &'a [ColumnName<'a>],
+        aggregates: &'a [(ColumnName<'a>, AggregateType)],
+    ) -> ReadGroupResults<'a> {
+        if !self.has_all_columns(&group_columns) {
+            return ReadGroupResults::default(); //TODO(edd): return an error here "group key column x not found"
```
[the remark this answers is not in the document]
Comment (e-dard) [reply segment_store/src/table.rs:183]: By the time any of this is hooked up it will all be implemented. I'll change all these to `todo!()` on the next PR though. Good idea 👍

File: segment_store/src/segment.rs [hunk 4]
```diff
@@ -496,6 +672,89 @@ impl<'a> std::fmt::Display for &ReadFilterResult<'a> {
...
+
+impl<'a> std::fmt::Display for &ReadGroupResult<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.is_empty() {
+            return Ok(());
+        }
+
+        let expected_rows = self.group_keys.len();
+        let mut row = 0;
+        while row < expected_rows {
```
[the remark this answers is not in the document]
Comment (e-dard) [reply segment_store/src/segment.rs]: Boy. I don't know what I was thinking here. Fixed in 44c05db. Good spot.

File: segment_store/src/segment.rs [hunk 5]
```diff
@@ -684,6 +943,82 @@ west,4
...
+north,GET,10
+south,PUT,203
+west,GET,100
+west,POST,304
+",
+            ),
+            (
+                vec![],
+                vec!["region", "method", "env"],
+                vec![("counter", AggregateType::Sum)],
```
[the remark this answers is not in the document]
Comment (e-dard) [reply segment_store/src/segment.rs:992]: 👍  See 0e77656

File: segment_store/src/column.rs [hunk 6]
```diff
@@ -1871,9 +2039,139 @@ pub enum AggregateResult<'a> {
     Last(Option<(i64, Value<'a>)>),
 }
 
+#[allow(unused_assignments)]
+impl<'a> AggregateResult<'a> {
+    pub fn update(&mut self, other: Value<'a>) {
```
[the remark this answers is not in the document]
Comment (e-dard) [reply segment_store/src/column.rs:2044]: The code in this PR is already doing vectorised aggregation because it's calling the aggregate functions directly on the columns. See here: https://github.com/influxdata/influxdb_iox/blob/0e77656527d45cfcee77eb210c6bad7e46a45bb6/segment_store/src/segment.rs#L451-L469

This code actually snuck in to this PR really because I don't need it until the general case, which is coming in this week. The reason it's dispatch per row is because typically the general group by case involves handling each row one at a time to build the group keys. In that case, performing non-vectorised aggregates probably won't have much if any perf downside. Of course I will be benchmarking to answer this question :-)

File: segment_store/src/column.rs [hunk 7]
```diff
@@ -1929,6 +2233,80 @@ impl Scalar {
     }
 }
 
+impl<'a> std::ops::AddAssign<&Scalar> for Scalar {
```
[the remark this answers is not in the document]
Comment (e-dard) [reply segment_store/src/column.rs:2236]: I agree. These implementations are not the best way to go if you can do vectorised aggregates. There are specialised methods on the column encodings for that.

**Comment by e-dard** [comment]
> I was somewhat confused about the explanation in the PR's description --
In your example, I can't tell if we are grouping by "region", or "region1", "region2" (aka are the two encoded columns the same column that needs to be aggregated, or are they two columns that form a compound group key-- I think it is the latter , and I found https://github.com/influxdata/influxdb_iox/pull/527/files#diff-05b254f51e0af46c018ef89da62237a52959aed4c59c6651b4f83835295fe51dR383-R413 to help understanding what is going on with this code better

It was the latter - it's about grouping on multiple columns. I have added more context/example to the PR description; hope it helps. The link doesn't seem to take me anywhere - I think linking to the code diff on the PR sometimes breaks. I have found if you click on "View File" on the PR diff it takes you somewhere you can link to.

File: segment_store/src/column.rs [hunk 8]
```diff
@@ -157,6 +168,27 @@ impl Column {
         }
     }
 
+    /// All logical values in the column.
+    pub fn all_values(&self) -> Values<'_> {
+        match &self {
+            Column::String(_, data) => data.all_values(),
```
Comment (domodwyer) [inline segment_store/src/column.rs:174]: I'm curious, is the explicit chunking in [all_values()](https://github.com/influxdata/influxdb_iox/blob/main/segment_store/src/column/fixed.rs#L147-L152) in order to coerce the rustc into vectorising it?

Comment (e-dard) [reply segment_store/src/column.rs:174]: Thanks for spotting this. Actually the chunking in the encoding implementations of `all_values` was done like that because I used it another spot to do some aggregation. The point was to avoid some extra bounds checks by doing `chunks[3]` then `[2]` `[1]` `[0]` etc. But for this I moved the order around (because the order matters). But I should have just reverted back to a simple loop because it's just as performant (I just benchmarked it). I'll tidy this up in another PR 👍

File: segment_store/src/lib.rs [hunk 9]
```diff
@@ -209,6 +209,26 @@ impl Store {
     }
 }
 
+/// Generate a predicate for the time range [from, to).
+pub fn time_range_predicate<'a>(from: i64, to: i64) -> Vec<segment::Predicate<'a>> {
```
Comment (domodwyer) [inline segment_store/src/lib.rs:213]: I couldn't see this being called - might be dead code?

We could also make use of [`Range`](https://doc.rust-lang.org/stable/std/ops/struct.Range.html) to encode these bounds - it has a well-defined half-open range, or even be generic over an impl of [`RangeBounds`](https://doc.rust-lang.org/stable/std/ops/trait.RangeBounds.html) that allows the caller to be specific with bounds handling (inc. unlimited)

You'd then be able to implement `From<Range<u64>> for Predicate` which would be [rather concise](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0d21dc9c06adbc8b15b6953daa46fa90):

```rust
    let time_range: Predicate = (start..end).into();
```

**Review by domodwyer — APPROVED** [review]
Looks neat - I'm not totally sure I fully the layout of grouped keys, etc yet though!

File: segment_store/src/lib.rs (hunk 9 above)
Comment (e-dard) [reply segment_store/src/lib.rs:213]: will address this cruft this week.

File: segment_store/src/column/dictionary.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alamb) [reply segment_store/src/column/dictionary.rs:761]: i see

**Comment by carols10cents** [comment]
Trying to make sure I understand the PR description first... I'm not sure I do, because I think there's a value missing in the intersection?

> we could simply intersect the bitsets for each group combination resulting in the final group keys and row_ids to apply the aggregates to:
> 
> (1, 1) -> {1, 3}
> (1, 2) -> {0}
> (2, 2) -> {5}
> ```

Shouldn't this also contain `(3, 3) -> {2}`?

**Comment by e-dard** [comment]
@carols10cents 🤦 fixing now. Let me know if you can review this as I was going to merge it today otherwise 😄

File: segment_store/src/column.rs [hunk 10]
```diff
@@ -1871,9 +1967,139 @@ pub enum AggregateResult<'a> {
...
+                    }
+                }
+                (Value::Scalar(_), Value::Null) => {} // do nothing
+                (Value::Scalar(a), Value::Scalar(b)) => {
+                    if a < b {
+                        *v = other;
+                    }
+                }
+                (_, _) => unreachable!("not a possible variant combination"),
+            },
```
Comment (carols10cents) [inline segment_store/src/column.rs:2125]: The code in here looks kind of like a partial `PartialOrd` implementation, and I see that you have `PartialOrd` derived on `Value`... do you know of a reason this logic can't be put in `PartialOrd` and then you can call `max` and `min` sorts of functions?

File: segment_store/src/column.rs [hunk 11]
```diff
@@ -1987,14 +2365,21 @@ pub enum Value<'a> {
 }
 
 impl Value<'_> {
-    fn scalar(&self) -> &Scalar {
+    pub fn is_null(&self) -> bool {
+        if let Self::Null = self {
+            return true;
+        }
+        false
+    }
```
Comment (carols10cents) [inline segment_store/src/column.rs:2373]: Can this use `matches!` like `Scalar::is_null` does?

File: segment_store/src/column.rs [hunk 12]
```diff
@@ -1929,6 +2233,80 @@ impl Scalar {
     }
 }
 
+impl<'a> std::ops::AddAssign<&Scalar> for Scalar {
+    fn add_assign(&mut self, _rhs: &Scalar) {
+        if _rhs.is_null() {
```
Comment (carols10cents) [inline segment_store/src/column.rs:2238]: Usually starting a parameter name with an underscore indicates it's unused in the body, but I see `_rhs` is immediately being used... Is there a reason you wanted this to be `_rhs` rather than `rhs`?

File: segment_store/src/column.rs [hunk 13]
```diff
@@ -1929,6 +2233,80 @@ impl Scalar {
...
+                    panic!("invalid AddAssign types");
+                };
+            }
+            _ => unimplemented!("unsupported and to be removed"),
+        }
+    }
+}
+
+impl<'a> std::ops::AddAssign<&Scalar> for &mut Scalar {
+    fn add_assign(&mut self, _rhs: &Scalar) {
```
Comment (carols10cents) [inline segment_store/src/column.rs:2271]: same `_rhs` here

File: segment_store/src/column.rs [hunk 14]
```diff
@@ -1929,6 +2233,80 @@ impl Scalar {
...
+                } else {
+                    panic!("invalid AddAssign types");
+                };
+            }
+            _ => unimplemented!("unsupported and to be removed"),
+        }
+    }
+}
+
+impl<'a> std::ops::AddAssign<&Scalar> for &mut Scalar {
```
Comment (carols10cents) [inline segment_store/src/column.rs:2270]: I see `'a` being declared here but I don't see it being used anywhere?

**Review by carols10cents — CHANGES_REQUESTED** [review]
Going to eat lunch but submitting what I have so far

File: segment_store/src/column.rs (hunk 10 above)
Comment (e-dard) [reply segment_store/src/column.rs:2125]: I feel like I tried to go down this path but it didn't work out. I'll see if I can get it working.

File: segment_store/src/column.rs (hunk 14 above)
Comment (e-dard) [reply segment_store/src/column.rs:2270]: surprised `clippy` didn't spot that.

Comment (carols10cents) [reply segment_store/src/column.rs:2270]: Submitted an issue: https://github.com/rust-lang/rust-clippy/issues/6437

File: segment_store/src/column.rs [hunk 15]
```diff
@@ -2253,71 +2660,79 @@ impl RowIDs {
...
-            (RowIDs::Bitmap(_self), RowIDs::Bitmap(ref other)) => _self.or_inplace(other),
+            (Self::Bitmap(_self), RowIDs::Bitmap(other)) => _self.or_inplace(other),
+            // N.B this seems very inefficient. It should only be used for testing.
+            (Self::Vector(_self), Self::Bitmap(other)) => {
+                let mut bm: Bitmap = _self.iter().cloned().collect();
+                bm.or_inplace(other);
+
+                _self.clear();
+                _self.extend(bm.iter());
+            }
```
Comment (carols10cents) [inline segment_store/src/column.rs:2735]: Another place with a variable starting with an underscore that is being used... I'd call this `inner` or similar? 

Also have, you benchmarked this Vector+Bitmap case to see if it's actually inefficient? I agree the code seems like it's doing a lot of work, but it's hard to be sure sometimes... if it's actually only relevant for testing anyway, it's fine.

File: segment_store/src/segment.rs [hunk 16]
```diff
@@ -129,11 +132,8 @@ impl Segment {
     }
 
     // Returns a reference to a column from the column name.
-    fn column_by_name(&self, name: ColumnName<'_>) -> Option<&Column> {
-        match self.all_columns_by_name.get(name) {
-            Some(&idx) => Some(&self.columns[idx]),
-            None => None,
-        }
+    fn column_by_name(&self, name: ColumnName<'_>) -> &Column {
+        &self.columns[*self.all_columns_by_name.get(name).unwrap()]
```
Comment (carols10cents) [inline segment_store/src/segment.rs:136]: Is this `unwrap` really ok to do? How can we be sure?

File: segment_store/src/segment.rs [hunk 17]
```diff
@@ -166,19 +166,19 @@ impl Segment {
     ///
     /// Right now, predicates are conjunctive (AND).
     pub fn read_filter<'a>(
-        &self,
+        &'a self,
         columns: &[ColumnName<'a>],
         predicates: &[Predicate<'_>],
-    ) -> ReadFilterResult<'a> {
+    ) -> ReadFilterResult<'_> {
```
Comment (carols10cents) [inline segment_store/src/segment.rs:172]: I have yet to see a use of `&'a self` that's actually what you want. When I take all the lifetimes out of here:

```
    pub fn read_filter(
        &self,
        columns: &[ColumnName<'_>],
        predicates: &[Predicate<'_>],
    ) -> ReadFilterResult<'_> {
```

I get this error message:

```
error[E0623]: lifetime mismatch
   --> segment_store/src/segment.rs:174:9
    |
170 |         columns: &[ColumnName<'_>],
    |                    -------------- this parameter and the return type are declared with different lifetimes...
171 |         predicates: &[Predicate<'_>],
172 |     ) -> ReadFilterResult<'_> {
    |          --------------------
173 |         let row_ids = self.row_ids_from_predicates(predicates);
174 |         ReadFilterResult(self.materialise_rows(columns, row_ids))
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...but data from `columns` is returned here

```

Which is using lifetime info from `materialize_rows`, which is also using `&'a self`, and if I take the lifetimes out of there:

```
    fn materialise_rows(
        &self,
        columns: &[ColumnName<'_>],
        row_ids: RowIDsOption,
    ) -> Vec<(ColumnName<'_>, Values<'_>)> {
```

now I see:

```
error[E0623]: lifetime mismatch
   --> segment_store/src/segment.rs:184:38
    |
179 |         columns: &[ColumnName<'_>],
    |                    -------------- this parameter and the return type are declared with different lifetimes...
180 |         row_ids: RowIDsOption,
181 |     ) -> Vec<(ColumnName<'_>, Values<'_>)> {
    |          ---------------------------------
...
184 |             RowIDsOption::None(_) => results, // nothing to materialise
    |                                      ^^^^^^^ ...but data from `columns` is returned here
```

I'm starting to think the problem here is with `ReadFilterResult`, its definition ties the lifetimes of the column names and the values together but I don't think that's the case. I'm going to try some things out locally and maybe send a branch.

**Review by carols10cents — CHANGES_REQUESTED** [review]
Pausing to try some things

File: segment_store/src/segment.rs (hunk 17 above)
Comment (e-dard) [reply segment_store/src/segment.rs:172]: I was going to ping you this week about the lifetimes on self. I remember you telling me they're often not what we want but I couldn't get rid of them this time. Let me know if you figure it out!

File: segment_store/src/column.rs (hunk 15 above)
Comment (e-dard) [reply segment_store/src/column.rs:2735]: yeah I did preface with *seems* given i didn't take the trouble to benchmark it. And the note will let me pause for thought if I do need to use it in a non-test path. Fixing the variable.

File: segment_store/src/segment.rs (hunk 16 above)
Comment (e-dard) [reply segment_store/src/segment.rs:136]: `unwrap` is OK because the segment is immutable and it's up to the segment owner (`table.rs`) to validate that the segment _may_ have data for the column first. If the segment couldn't have data for the column (because the column didn't exist, say) then the query wouldn't make it this far. I'll add a comment.

**Review by carols10cents — APPROVED** [review]
I'm good with this, we're going to keep improving on it in future PRs :)

**Comment by alamb** [comment]
🎉

**Comment by e-dard** [comment]
@carols10cents has some sweet refactoring to clarify lifetimes in the segment store. We will be getting that merged in a separate PR next.
